Request broadcast data after Overlay is loaded and simplify EntryListTracker

iFuSiiOnzZ opened the pull request:
* If you started the application while you were in a session, the broacast data could not be accessed as the broadcast system had already made all the relevant notifications. This has been temporarily solved by asking broadcast system data again.

* Unify EntryListTracker data processing. This way there is only one point in charge of updating the data allowing a better maintenance, adding new functionalities, etc.

iFuSiiOnzZ committed:
Simplify EntryListTracker

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Add hack to request the broadcast data when the overlay is already loaded

iFuSiiOnzZ committed:
Add debug information for loaded track map

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Remove duplicate cases for brodcast event handler

iFuSiiOnzZ committed:
Track length is now accessed through the broadcast api

RiddleTime commented on Race_Element.HUD.ACC/Overlays/Driving/TrackMap/TrackMapOverlay.cs:
```diff
@@ -185,6 +181,7 @@ private void OnMapPositionsCallback(object sender, List<TrackPoint> positions)
         _trackPositions = track;
 
         _mapCache.Map = TrackMapDrawer.CreateLineFromPoints(_config.Colors.Map, _config.General.Thickness, _margin, _trackPositions, _trackBoundingBox);
+        Debug.WriteLine("[MAP] " + pageStatic.Track.ToLower() + " -> [" + _scale + "] [" + broadCastTrackData.TrackMeters + "] [" + _trackPositions.Count + "]");
```
Try using this when formatting values into a string:

```csharp
float max = 30.372348;
Debug.WriteLine(`$"some text: {max:F2}");

```
`some text: 30.37`

RiddleTime replied:
also hiding the the track name (capital letter usage), could obfuscate reasons why developers of games made certain choices, I'd not hide that

iFuSiiOnzZ replied:
I put it to lower to normalize the names, as some tracks are upper case and some lower case.

RiddleTime commented:
Would be an idea to have the broadcast UDP information and or status displayed as huds like the shared memory ones are currently are?

iFuSiiOnzZ commented:
Why kind of information? Apart from "TrackData" which is set once per session and "ConnectionState", all the other stuff is related to the cars on track (updated every X milliseconds). Well, probably "CarInfo" should be also quite stable (until a player gets into session or out of the session).

iFuSiiOnzZ committed:
Use C# native string format

iFuSiiOnzZ committed:
Check for null before calling the function

iFuSiiOnzZ committed:
Break map render into small function

iFuSiiOnzZ committed:
Simplify ShouldRender

iFuSiiOnzZ committed:
Change if statement

RiddleTime commented:
> Why kind of information? Apart from "TrackData" which is set once per session and "ConnectionState", all the other stuff is related to the cars on track (updated every X milliseconds). Well, probably "CarInfo" should be also quite stable (until a player gets into session or out of the session).

Perhaps just the connection and maybe the local car if available, though I could expand on the already existing broadcast realtime hud (pitwall section).

iFuSiiOnzZ commented:
> Perhaps just the connection and maybe the local car if available, though I could expand on the already existing broadcast realtime hud (pitwall section).

You could see the kilometers on track, see if it had any incident (last incident for instance), when it gets out of the pit have en extra state that says "Out lap", etc, but I have no idea, probably someone who does live broadcast as a better idea of what is useful and what not.

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

RiddleTime committed:
change entry list dictionary to concurrent dictionary

Changed _entryListCars to a ConcurrentDictionary, this is a threadsafe collection.

RiddleTime committed:
Update EntryListTracker.cs

removed last locks

RiddleTime committed:
seal classes for extra performance

RiddleTime committed:
broadcasttracker: prevent spamming of request data

add 5 second debounce to prevent multiple AbstractOverlays from all requesting new data.

RiddleTime committed:
Merge branch 'dev' into pr/184

RiddleTime committed:
Merge branch 'dev' into pr/184

RiddleTime committed:
Merge branch 'dev' into pr/184

RiddleTime committed:
Merge branch 'dev' into pr/184

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

iFuSiiOnzZ committed:
Merge branch 'RiddleTime:dev' into dev

RiddleTime commented:
I gave the debounce (5 second break) for the request of entry list a shot, that seems to work. As long as the generic data system is not finished this cannot be solved properly. With a generic system we just need to keep track of the data in 1 place and make it available like the shared memory, then additionaly we can have events that trigger like lap completed.
